lib/HLSL/DxilCondenseResources.cpp

```diff
@@ -2781,12 +2787,397 @@ ModulePass *llvm::createDxilAllocateResourcesForLibPass() {
+          Sel->replaceAllUsesWith(newCI);
+          CleanupInsts.insert(Sel);
+          // put newCI in HandleToArgs.
+          HandleToArgs[newCI] = args;
+        } else {
+          newSel->eraseFromParent();
+        }
+      } else {
+        DXASSERT(false, "otherwise, non-select/phi in HandleSelects set");
+      }
+    }
+  }
+
+  void CollectResources(DxilModule &DM) {
+    ValueSetVector tmpHandleSelects;
+    hlsl::OP *hlslOP = DM.GetOP();
+    if (hlslOP->IsDxilOpUsed(DXIL::OpCode::CreateHandleFromHeap)) {
+      Function *F = hlslOP->GetOpFunc(DXIL::OpCode::CreateHandleFromHeap,
+                                      Type::getVoidTy(DM.GetCtx()));
+      for (User *U : F->users()) {
+        DxilInst_CreateHandleFromHeap Hdl(cast<CallInst>(U));
+        HandleToArgs[U] = {Hdl.get_index(), Hdl.get_samplerHeap_val(),
+                           Hdl.get_nonUniformIndex_val(), true};
+        for (User *HdlU : U->users()) {
+          if (isa<PHINode>(HdlU) || isa<SelectInst>(HdlU)) {
+            tmpHandleSelects.insert(HdlU);
+          }
+        }
+      }
+    }
+    // Collect phi/sel of other phi/sel selected handles.
+    while (!tmpHandleSelects.empty()) {
+      HandleSelects.insert(tmpHandleSelects.begin(), tmpHandleSelects.end());
+      ValueSetVector newHandleSelects;
+      for (Value *Hdl : tmpHandleSelects) {
+        for (User *HdlU : Hdl->users()) {
+          if (HandleSelects.count(HdlU))
+            continue;
+          if (isa<PHINode>(HdlU) || isa<SelectInst>(HdlU)) {
+            newHandleSelects.insert(HdlU);
+          }
+        }
+      }
+      tmpHandleSelects = newHandleSelects;
+    }
+  }
+
+  void DoTransform(hlsl::OP *hlslOP) {
+    CreateSelectsForHandleSelects();
+    PropagateHeapArgs();
+    UpdateSelectsForHandleSelect(hlslOP);
+    CleanupUnusedValues(CleanupInsts);
+  }
+
+  bool runOnModule(llvm::Module &M) {
+    DxilModule &DM = M.GetOrCreateDxilModule();
+    hlsl::OP *hlslOP = DM.GetOP();
+
+    CollectResources(DM);
+
+    // If no selects or allocas are involved, there isn't anything to do
+    if (HandleSelects.empty())
+      return false;
+
+    DoTransform(hlslOP);
+
+    return true;
+  }
+};
+
+namespace {
+// Make sure no phi/sel on annotateHandle.
+bool sinkAnnotateHandleAfterSelect(DxilModule &DM, Module &M) {
+  // Collect AnnotateHandle calls.
+  SmallVector<CallInst *, 4> annotHdls;
+  hlsl::OP *op = DM.GetOP();
+  LLVMContext &Ctx = M.getContext();
+  Type *pVoidTy = Type::getVoidTy(Ctx);
+  Function *annotHdlFn = op->GetOpFunc(DXIL::OpCode::AnnotateHandle, pVoidTy);
+  for (auto it : op->GetOpFuncList(OP::OpCode::AnnotateHandle)) {
+    Function *F = it.second;
+    if (F == nullptr)
+      continue;
+    for (auto U = F->user_begin(); U != F->user_end();) {
+      CallInst *CI = dyn_cast<CallInst>(*(U++));
+      annotHdls.emplace_back(CI);
+    }
+  }
+  if (annotHdls.empty())
+    return false;
+
+  DenseSet<Instruction *> selectAnnotHdls;
```

Comment:
Need to use SetVector, otherwise iteration will be pointer-dependent, leading to potentially different orderings of replacement calls after phi nodes on different runs (nondeterminism).